### PR TITLE
odb: add polygon support

### PR DESCRIPTION
**Pull request by gadfort**
Adds:
- Add `Polygon` geom class (pretty simple improvements can be made in the future to handle intersections, etc as needed).
- add `dbPBox` to hold polygons (currently enabled for `dbMaster` (for obstructions) and `dbMPin` (for pin geometries))
  - this class hold a set of `dbBox`, which represents the decomposed version of the polygon
- `dbBoxItr` was modified to iterate through the polygon boxes to maintain backwards compatibility.
- `dbPBoxItr` a `dbPBox` iterator.
- `getPolygon??` and `getNonPolygon??` for tools to be able to access the polygons and non-polygon boxes separately.

Functionally, there should be no changes, except that the dbBox ordering may be slightly different due to the polygon iteration.

GUI modifications:
- adds support for the gui to paint the polygon shapes
- adds control to be paint either polygon or decomposed boxes (for debugging)

Previous gui view:
Screenshot 2024-08-04 105229

New gui view:
Screenshot 2024-08-04 105218

Future work:
- enable support in lef writer
- enable support in PDN, PAD, and PSM

### CODE REVIEW

**Comment by gadfort** [comment]
@maliberty in doing this I realized we need a better way to pass around the GUI state information since the `std::function`'s are pretty clumsy. For now I kept the pattern, but after this it might be worth creating a `GUIState` class to hold this information for us.

File: src/odb/src/db/dbPBoxItr.h [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ///////////////////////////////////////////////////////////////////////////////
 // BSD 3-Clause License
 //
-// Copyright (c) 2019, Nefelus Inc
+// Copyright (c) 2024, The Regents of the University of California
```
Comment (QuantamHD) [inline src/odb/src/db/dbPBoxItr.h:4]: This is not correct. You can't eliminate others copyright. This should read

Copyright (c) 2019-2024, Nefelus Inc, Zero ASIC Corporation

Also other areas where you updated the copyright.

Comment (gadfort) [reply src/odb/src/db/dbPBoxItr.h:4]: That was a copy paste oversight on my part. This and the dbPBox files are new.

File: src/odb/src/codeGenerator/schema/chip/dbPBox.json
Comment (maliberty) [inline src/odb/src/codeGenerator/schema/chip/dbPBox.json:1]: Rename to dbPolygon

File: src/odb/include/odb/db.h [hunk 2]
```diff
@@ -5828,6 +5844,16 @@ class dbMPin : public dbObject
...
 
+  ///
+  /// Get the polygon geometry of this pin.
+  ///
+  dbSet<dbPBox> getPolygonGeometry();
+
+  ///
+  /// Get the non-polygon geometry of this pin.
+  ///
+  dbSet<dbBox> getNonPolygonGeometry();
```
Comment (maliberty) [inline src/odb/include/odb/db.h:5855]: Replace with an option on getGeometry

File: src/gui/src/mainWindow.cpp [hunk 3]
```diff
@@ -593,6 +598,10 @@ void MainWindow::createActions()
   show_dbu_->setCheckable(true);
   show_dbu_->setChecked(false);
 
+  show_db_view_ = new QAction("Show DB view", this);
```
Comment (maliberty) [inline src/gui/src/mainWindow.cpp:601]: Show polygon decomposition.  Preferably not a menu choice as it only useful to developer.

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/odb/src/codeGenerator/schema/chip/dbPBox.json [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      "flags":["no-set"]
+    },
+    {
+      "name":"design_rule_width_",
+      "type":"int",
+      "flags":["no-set"]
+    },
+    {
+      "name": "owner_",
+      "type": "uint",
```
[the remark this answers is not in the document]
Comment (maliberty) [reply src/odb/src/codeGenerator/schema/chip/dbPBox.json:33]: Multiple possible owners

**Comment by gadfort** [comment]
@maliberty this segfault seems unrelated:
```
repair_timing -skip_gate_cloning
Signal 11 received
Stack trace:
 0# 0x00006300867A59BB in /tmp/workspace/OpenROAD-Public_PR-5504-head/build/src/openroad
 1# 0x000079A6DA841520 in /lib/x86_64-linux-gnu/libc.so.6
 2# sta::VertexPathIterator::VertexPathIterator(sta::Vertex*, sta::RiseFall const*, sta::PathAnalysisPt const*, sta::StaState const*) in /tmp/workspace/OpenROAD-Public_PR-5504-head/build/src/openroad
 3# sta::CheckCrpr::maxCrpr(sta::ClkInfo*) in /tmp/workspace/OpenROAD-Public_PR-5504-head/build/src/openroad
 4# sta::ArrivalVisitor::pruneCrprArrivals() in /tmp/workspace/OpenROAD-Public_PR-5504-head/build/src/openroad
 5# sta::ArrivalVisitor::visit(sta::Vertex*) in /tmp/workspace/OpenROAD-Public_PR-5504-head/build/src/openroad
 6# 0x0000630086AE3EFD in /tmp/workspace/OpenROAD-Public_PR-5504-head/build/src/openroad
 7# sta::DispatchQueue::dispatch_thread_handler(unsigned long) in /tmp/workspace/OpenROAD-Public_PR-5504-head/build/src/openroad
 8# 0x000079A6DAC0D253 in /lib/x86_64-linux-gnu/libstdc++.so.6
 9# 0x000079A6DA893AC3 in /lib/x86_64-linux-gnu/libc.so.6
10# 0x000079A6DA925850 in /lib/x86_64-linux-gnu/libc.so.6
ERROR segmentation violation
```

**Review by maliberty — APPROVED** [review]

**Pull request merged**